### ISSUE TITLE
Custom transaction begin statement

### ISSUE THREAD

**Issue by JonasProgrammer** [yandex/ozo#156]
Hi,

I have a question regarding `ozo::begin()`: Currently the executed seems to be hardcoded to `BEGIN`. Is there any customization point for this? For now I simply copied the definition of `begin` and `begin_op` with a new name, but this is not really a solution.

Specifically, I need certain transactions to be `SERIALIZABLE`, whilst others are `READONLY`. While `SET` can do this as well, I don't see the point in sending two commands to the server when one would do.

If there is no such customization point, I'd be happy to help with a PR.

Thanks in advance

**Comment by thed636** [comment]
Hi Jonas!

> I have a question regarding ozo::begin(): Currently the executed seems to be hardcoded to BEGIN. Is there any customization point for this?

Not yet, but should be made. I have The Idea :) but no time at all to implement it. :(

> Currently I simply copied the definition of begin and begin_op with a new name, but this is not really a solution.

Absolutely.

> If there is no such customization point, I'd be happy to help with a PR.

Yes, it would be very helpful.

And here is **The Idea** :)

```c++
// Initially it was enum, but it is not easy to dispatch enum items comparing to types
struct isolation_level {
    constexpr struct serealizable_t{} serializable;
    constexpr struct repeatable_read_t{} repeatable_read;
    constexpr struct read_committed_t{} read_committed;
    constexpr struct read_uncommitted_t{} read_uncommitted;
};

struct transaction_mode {
    constexpr struct read_write_t read_write;
    constexpr struct read_only_t read_only;
};

template <class V>
struct deferrable_mode : V {
    constexpr auto operator !() const { 
        return deferrable_mode<typename std::negation<V>::type>{};
    }
};

constexpr deferrable_mode<std::true_type> deferrable; 
```

Here are some possible variants of usage

```c++
auto transaction = begin.with_options(isolation_level::serializable, deferrable)(connection_provider, time_constraint, yield);
```

```c++
const auto options = transaction_options(isolation_level:: serializable, deferrable);
auto transaction = begin(connection_provider, time_constraint, options, yield);
```

Maybe the options could participate in `transaction` type to provide type-safe transactions expectations. So the respective functions could be provided.

```c++
template <typename T>
constexpr auto get_isolation_level (const T&);
template <typename T>
constexpr auto get_transaction_mode (const T&);
template <typename T>
constexpr auto get_deferrable_mode (const T&);
```

So feel free to implement it or provide your own ideas.

**Comment by JonasProgrammer** [comment]
Hi Sergei,
glad to know this is indeed an API issue (and not me missing parts of the documentation again) and you already put some thoughts into this.

Your plans on this are a little more elaborated than I expected (I simply thought about a customization point for the statement), but I'll try and take the challenge. Should be a good meta programming exercise to cure me from Java disease.

That being said, please don't hold your breath on this for now, I likely cannot start working on this before next week.

**Comment by JonasProgrammer** [comment]
Ok, I have used your proposed interface and have something builder-like that generates proper queries based on what options tags are passed in, which could in turn be used to used to tag the transaction object itself and implement the proposed `get_*` functions.

Before getting this ready for a PR, I still have some questions regarding the structure:
* What do you consider `detail` and what `impl` (namespaces)? From the current code, I'd just guess that the `begin_option_builder` is a detail and the stuff it uses like `isolation_level_sql` belongs in `impl`, however I'm everything but sure.
* Do you have something like a `.clang-format` or `.editorconfig` for this project by any chance? Not that it's necessary, but I'd much rather run an automated tool against this than trying to mimic the current code style in CLion's settings and manually fixing what still looks different.

**Comment by thed636** [comment]
Hi, Jonas!

> Ok, I have used your proposed interface and have something builder-like that generates proper queries based on what options tags are passed in, which could in turn be used to used to tag the transaction object itself and implement the proposed `get_*` functions.

Wow, I'm really happy about this!

> Before getting this ready for a PR, I still have some questions regarding the structure:
> 
> * What do you consider `detail` and what `impl` (namespaces)? From the current code, I'd just guess that the `begin_option_builder` is a detail and the stuff it uses like `isolation_level_sql` belongs in `impl`, however I'm everything but sure.

Well, it's a very good question. :) Like the Universe we had a little of chaos at the beginning. Nowadays looks like there is no difference between `impl` and `detail` namespace. The `impl` namespace surely should be merged into `detail` namespace in the future. So it is better to use `detail` namespace for internal entities.

> * Do you have something like a `.clang-format` or `.editorconfig` for this project by any chance? Not that it's necessary, but I'd much rather run an automated tool against this than trying to mimic the current code style in CLion's settings and manually fixing what still looks different.

Sorry, but we didn't do any of these. :( And I slightly regret about `.clang-format` missing, because it should be done at the start of the project. On the over side - you need to achieve an agreement in code-style between several smart people with different opinions and this is a real pain, so I decided to have some freedom in code style. Sorry for such inconvenient. Let it be a price of freedom and peace. :)

**Comment by thed636** [comment]
@JonasProgrammer 
Hi Jonas!

Any news on this issue? Are you still interested?

**Comment by JonasProgrammer** [comment]
Hi Sergei!

Sorry for the delay, I frankly forgot about this after not looking at it for a week after your last reply and not being at home the week after that.

That being said, yes, I still am interested, but looking at what you added in `ozo/core/options.h`, I figured scraping the eye cancer that was my options builder and instead starting over with what is available now is the better way.

I'll try and finish it this time without forgetting about it ;)

One problem I did encounter, however, was integration testing: While it's fairly easy to validate the constructed begin statements are syntactically correct, there seems to be no way in PostgreSQL to query what parameters the current transaction was started with.
If unit-testing the statement generation and doing integration tests in terms that the statement does not lead to a syntax error is enough for you, everything is fine, but if you want the actual transaction state verified, I have no clue on how to do it.
